### PR TITLE
Sources Socket Handles

### PR DESCRIPTION
**Pull request by AustinShalit**
This solves #269.  

Maybe this method should be rewritten to use the Source super class more.

### CODE REVIEW

**Comment by ThomasJClark** [comment]
Nice catch, I totally forgot that `Socket` has `@Subscribe` methods.

Maybe it would be more future-proof to just have `eventBus.register(this);` in the constructor for `Socket`.  That way, we don't have to always remember to register new sockets

**Comment by JLLeitschuh** [comment]
No, bad idea. See #155

**Comment by AustinShalit** [comment]
How do we want to proceed?  I recommend we merge this PR for the 1.0.0 release and open a new issue to fix it for 1.0.1

**Comment by JLLeitschuh** [comment]
The long term term fix will be that the sockets to be constructed using dependency injection.
This fix should work for now.

**Comment by ThomasJClark** [comment]
> No, bad idea. See #155

That doesn't apply here because `Socket` subscribes to `ConnectionAddedEvent` and `ConnectionRemovedEvent`, which can't be sent before the socket is completely constructed (whoever posts  the event needs a reference to the socket to create the event)

**Comment by JLLeitschuh** [comment]
Alternatively, we could make it so that the Connection calls `addConnection` on the `Socket` when `Connection` receives a `ConnectionAddedEvent` and then, that way, sockets don't even have to be registered on the event bus.

**Comment by JLLeitschuh** [comment]
I think this is an appropriate way of handling this problem for the time being.
Do you agree @ThomasJClark or do you think it should be changed.

**Comment by ThomasJClark** [comment]
> I think this is an appropriate way of handling this problem for the time being.

Does "this" mean the way this PR does it, or the way you proposed? (no on the first, yes on second)

**Comment by JLLeitschuh** [comment]
Okay, fine I meant the first way.
@AustinShalit do you think you can resolve this the way I described?
Removing the subscribe methods from the socket and adding a `addConnection` and `removeConnection` method in their place?
If you can't get to this then just say so and I can take care of it.

**Comment by AustinShalit** [comment]
@JLLeitschuh Yea.  I will take care of this

File: core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
 
     private final SocketHint<Mat> imageOutputHint = SocketHints.Inputs.createMatSocketHint("Image", true);
     private final OutputSocket<Mat> outputSocket;
+    private final EventBus eventBus;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java]: This should no longer be necessary

File: core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java [hunk 2]
```diff
@@ -88,6 +89,7 @@ private MultiImageFileSource(
             final String[] paths,
             final int index) {
         super(exceptionWitnessFactory);
+        this.eventBus = checkNotNull(eventBus, "Event Bus was null.");
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java]: This too.

File: core/src/main/java/edu/wpi/grip/core/Socket.java [hunk 3]
```diff
@@ -134,26 +134,25 @@ public Direction getDirection() {
...
 
-    @Subscribe
-    public void onConnectionRemoved(ConnectionRemovedEvent event) {
-        if (event.getConnection().getInputSocket() == this || event.getConnection().getOutputSocket() == this) {
-            this.connections.remove(event.getConnection());
+    /**
+     * @param connection The connection to remove from this socket.
+     */
+    public void removeConnection(Connection connection) {
+        this.connections.remove(connection);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Socket.java]: Add a `checkNotNull(connection, "Can not remove null connection")`

File: core/src/main/java/edu/wpi/grip/core/Socket.java [hunk 4]
```diff
@@ -134,26 +134,25 @@ public Direction getDirection() {
         return ImmutableSet.copyOf(this.connections);
     }
 
+    /**
+     * @param connection The connection to add to this socket.
+     */
+    public void addConnection(Connection connection) {
+        this.connections.add(connection);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Socket.java]: `checkNotNull(connection, "Can not add null connection")`

File: core/src/main/java/edu/wpi/grip/core/Connection.java [hunk 5]
```diff
@@ -64,14 +66,14 @@ public void removeConnection(StepRemovedEvent e) {
...
-                this.eventBus.post(new ConnectionRemovedEvent(this));
+                socket.removeConnection(this);
                 return;
             }
         }
 
         for (Socket socket : e.getStep().getInputSockets()) {
             if (socket == this.inputSocket || socket == this.outputSocket) {
-                this.eventBus.post(new ConnectionRemovedEvent(this));
+                socket.removeConnection(this);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Yes! This is exactly right!

File: core/src/main/java/edu/wpi/grip/core/Connection.java [hunk 6]
```diff
@@ -49,6 +49,8 @@
     public void onConnectionAdded(ConnectionAddedEvent event) {
         if (event.getConnection().equals(this)) {
             inputSocket.setValueOptional(outputSocket.getValue());
+            inputSocket.addConnection(this);
+            outputSocket.addConnection(this);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Awesome!

File: core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java [hunk 7]
```diff
@@ -22,6 +21,7 @@
 import java.util.stream.Collectors;
 
 import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkNotNull;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java]: This shouldn't need to be here.

**Pull request merged**